### ISSUE TITLE
Commit Search-ability and Navigation

### ISSUE THREAD

**Issue by JosephTLyons** [desktop/desktop#7022]
**Please describe the problem you think should be solved**
Currently, other than commit messages themselves, there's no easy way to quickly differentiate between commits.  

Let me illustrate with an example.

=====

Two workers are building a piece of software and are communicating through some voice app (Discord, Slack, etc.).  They are currently reviewing some code, through GitHub Desktop, pushed by one of them the previous night.  

Person A: "Hey can you review this commit?"
Person B: "Which one?"
Person A: "The one that says Update-File-1.cpp"
Person B: "There's 5 different commits with that message, which one specifically?"
Person A: "The third one of those messages, it's about..... maybe 15 down from the latest commit?"
Person B: "Okay one second" * 1.. 2.. 3... ... *
Person B: "Okay I found it"

=====

The root of this issue is poorly written Git commit messages.  If everyone in the world wrote uniquely identifiable commit messages that accurately describe the change that was made, this wouldn't be an issue.  However, we cannot change the world and the bad habits of each and every developer.  What we do have control over is whether or not the software they are using as in interface to Git tries to remedy these issues.

I think it would be great if each "slice" (for lack of the proper term) in the history tab had an associated git commit number.  This could be the actual commit hash or just a simple integer that starts from 1 and counts up to the total commit number, with the latest commit having the total commit number that we see on GitHub's website:

<img width="993" alt="screen shot 2019-03-06 at 2 59 11 pm" src="https://user-images.githubusercontent.com/19867440/53909957-c59c5980-4020-11e9-86c0-d001845aa237.png">

We currently have git commit hashes included in the actual diff view:

<img width="938" alt="screen shot 2019-03-06 at 3 03 41 pm" src="https://user-images.githubusercontent.com/19867440/53910156-2deb3b00-4021-11e9-9e6d-4ef12a6868a3.png">

This is handy, but requires us to open each commit's diff view, so this isn't a solution for this specific problem.

So I propose implementing a system of unique references directly in the history view, on each "slice," as shown in here:

<img width="354" alt="screen shot 2019-03-06 at 4 15 50 am" src="https://user-images.githubusercontent.com/19867440/53910291-7f93c580-4021-11e9-8a62-a663772ae7e9.png">

**Comment by JosephTLyons** [comment]
@billygriffin , as per your request.

**Comment by ipkpjersi** [comment]
I'd love to see this. You provided a really clear point of view about why this feature would be good to have and I agree, this does seem really useful to have. Good idea.

**Comment by billygriffin** [comment]
Thanks @JosephTLyons, really appreciate the detailed write up! ❤️  If others have felt this pain as well, I'd love to hear how it manifests for you and your current workarounds today.

**Comment by ipkpjersi** [comment]
I just saw today that this issue referenced a search in history feature where you can enter text to search for. I think that should possibly be added to this issue/feature request as long as it's not super out-of-scope. Being able to refer to search commits for text (I remember I changed Feature X in File Y, so I search for Feature X File Y and I find the commit) then I am able to refer to that commit and tell my teammate that "hey, can you check out commit number 335 in repo main-site? It has a good example of how to do Thing Z". To me, being able to refer to commits by numbers is much more human-friendly than checksums which are inherently machine-friendly but not necessarily human-friendly. I think something that makes this program more human-friendly is going to be a good idea, and I think the suggestions proposed in this thread and in my current comment are a very good idea. It would make GitHub desktop a lot more useful IMO. Does anyone have any thoughts on this?

**Comment by JosephTLyons** [comment]
With the release of version 2.1.0, we can now search for text within the diff views.  This is extremely awesome; it adds speed to the Desktop workflow.  However, now that we can search within a single diff, I'm wanting the ability to search through all commits even more.  

This strays away from what I originally posted about with a commit number in each commit slice (although I still think that that would help with communication) and travels down the path of having a search bar for the git commit history pane.  I realize that this is probably not an easy thing to implement.  What all criteria should be searchable?  Should you only be able to search commit messages, or authors, or actual diff content?  I'm not sure how far that it should go, but I do think it could be released incrementally.  At first, it could just support maybe searching commit messages, but then add more search options as time goes on.  I think Sublime Merge's search function is extremely exhaustive.  It can do pretty much anything.  That could be a source of inspiration for Desktop, but I don't think it needs to be that extensive.

**Comment by Seb33300** [comment]
The https://github.com/desktop/desktop/issues/1543 issue was close in favor of this one.

On my side I am waiting for the ability to search in the commit message history for a long time.
And I had also the same remark by almost all of my colleagues wondering why this is not a searchbox:

![image](https://user-images.githubusercontent.com/915273/81770528-bfe6b180-950a-11ea-8f19-ccda37b26be8.png)

**Comment by erlandl4g** [comment]
Seb - you are not alone waiting for the Search ability.

**Comment by Ali-00-Ridha** [comment]
will this feature ever come to life?

**Comment by attheshow** [comment]
I'm just trying out the app this week for the first time. I used to use Gitbox for the longest time but it hasn't been working for me in the latest OS. It had a great search capability so I'm quite surprised to find that Github Desktop is missing this feature.

**Comment by steveward** [comment]
From #13396:

> ### Describe the feature or problem you’d like to solve
> 
> It's been a really long time since I've wanted the file search feature in GitHub Desktop! In large projects like huge Web/Mobile Applications, it becomes really hard to keep track of what files were updated/added/removed in a specific commit of a branch. The file search feature will make it very simple to find updated files and show their changes as opposed to finding the files manually by scrolling through the large list of updated files!
> 
> ### Proposed solution
> The proposed solution is to add a `<FancyTextbox/>`  right above the file list, like the one used in "search for branches to compare" in GitHub Desktop. This would be a huge productivity booster for many developers who would also want to search for a specific file on other branch commits!
> 
> ****_I would also like to ask if I could start working on this feature? I would find it really interesting to contribute and introduce this feature in the Desktop app._**

**Comment by Seb33300** [comment]
This issue is about searching in the commit message and comment, not by file name

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/14643

**Comment by subspacefield** [comment]
This was entered more than 3 years ago and still no traction? I had to load Sourcetree so I could search commit history on a development branch.

**Comment by attheshow2** [comment]
Lately I've been using Github Desktop for everything but searching. For searching I have to temporarily switch to Tower. Once I've found what I need, it's back to Github Desktop. 😊

**Comment by mr21** [comment]
[Seb33300](https://github.com/desktop/desktop/issues/7022#issuecomment-627734015), i was thinking the same


Please GitHubDesktop we are begging 🙏🙏🙏
![image](https://user-images.githubusercontent.com/850754/186162582-f614db40-6592-4220-9a86-5abcc0baddb9.png)

And super nice work since then btw ❤️

**Comment by a2937** [comment]
Some sort of history searching would be absolutely amazing in locating commits.  It would also make remembering when someone did what a lot smoother. I'm of the school of thought that the commit message itself should be used for searching as the numbers and letters are useless to people who don't know how they are obtained.

**Comment by sfeinstein-goodrx** [comment]
Search is a core feature, please add it

**Comment by pedropsi** [comment]
It would be really useful to be able to search **at least the commit titles**, especially to gather relevant commits quickly, e.g. ahead of a squash operation.

Here's a mock-up of how it would look like:

![image](https://user-images.githubusercontent.com/36260325/196363324-fc520c22-6c60-456e-b2f1-3a26c7f325a0.png)


At the same time, the "branch compare" option would move to inside the "branches" arch-tab, as that is the most natural place for multi-branch-related operations. By reusing the search box already in place there, the whole interface would (probably) become more coherent and intuitive.

![image](https://user-images.githubusercontent.com/36260325/196369757-76a03e1a-06e2-48ef-bad1-1c7bfac4bc79.png)

**Comment by meirroth** [comment]
+1 on this.

**Comment by Andrej730** [comment]
Looking for that feature too, currently I end up just going to github webpage and search for commits.

**Comment by attheshow** [comment]
Apparently this is a feature that was worked on at one point but it conflicted with some other features so it hasn't been included in the app yet and is on hold.

https://github.com/desktop/desktop/tree/search-commits-by-text

**Comment by BinaryAlan** [comment]
It is difficult to understand why this problem has not been resolved after five or six years. Since people need this feature, it should be added.

**Comment by a2937** [comment]
Until staff gives this feature further design consideration , it won't be added. Though I would _love_ to implement this myself.

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/17575.

**Comment by sergiou87** [comment]
Requested in https://github.com/desktop/desktop/issues/17621 too

> Add new block below selecting branch to compare for filters, Apply filters on history tab by: 
> 
> 1. Author
> 2. Dates (Allow Range)
> 3. Phrase

**Comment by tidy-dev** [comment]
Requested in #17759

**Comment by 0xdevalias** [comment]
RE: https://github.com/desktop/desktop/issues/7022#issue-417986356

> This could be the actual commit hash or just a simple integer that starts from 1 and counts up to the total commit number, with the latest commit having the total commit number that we see on GitHub's website:

Personally I think using the short commit hash is the better choice here, as it sticks with well known/well used git patterns; and therefore will translate well even if one user is using GitHub Desktop, and another GitHub web, and another just plain old `git`; the same hash/reference will work for all of them.

---

RE: https://github.com/desktop/desktop/issues/7022#issuecomment-507927519

> With the release of version 2.1.0, we can now search for text within the diff views.

I had no idea that was even possible! Pretty neat. Not super discoverable, as I don't think I have seen anything in the UI that exposes this; it just happens to work if you press Command+F with the diff view open. I would have expected that it would also have been shown in the settings dropdown on the diff or similar. Maybe even just a little magnifying glass next to that, though that might clutter the UI too much for a feature i've rarely had a need to use.

> This strays away from what I originally posted about with a commit number in each commit slice (although I still think that that would help with communication) and travels down the path of having a search bar for the git commit history pane. I realize that this is probably not an easy thing to implement. What all criteria should be searchable? Should you only be able to search commit messages, or authors, or actual diff content? I'm not sure how far that it should go, but I do think it could be released incrementally. At first, it could just support maybe searching commit messages, but then add more search options as time goes on.
 
I found my way to this issue wanting to search/filter just the commit messages themselves. Occasionally I could see value in it being able to search on other fields/diff content as well, but majority of the time if I wanted to do something that deep I would fall back to doing it on the command line.

As one example, in my dotfiles I am often referring back to my commit history to see/copy the commit messsage style/pattern I used when editing a file in a certain way; and currently I either need to manually scroll through the list to find it, or drop back to `git` command line for it. Being able to quickly filter the commits by commit message would make this a lot easier.

When I first saw the 'select branch to compare' textbox, (ignoring the text) my first expectation of that feature was that it would search/filter the commit messages.

---

RE: https://github.com/desktop/desktop/issues/7022#issuecomment-627734015

> And I had also the same remark by almost all of my colleagues wondering why this is not a searchbox

Agreed. My default expectation based on common UI/UX patterns is that the current 'branch compare' search box would make a lot more sense as a 'search commits' filter.

---

RE: https://github.com/desktop/desktop/issues/7022#issuecomment-1281985382

> It would be really useful to be able to search at least the commit titles, especially to gather relevant commits quickly, e.g. ahead of a squash operation.

> At the same time, the "branch compare" option would move to inside the "branches" arch-tab, as that is the most natural place for multi-branch-related operations.

While I'm not 100% sure I like the mockup for this, I definitely feel like the 'branch' section is the correct place for this sort of 'select compare branch' feature.

Though another alternative, if the current UI patterns didn't want to be changed too much:

- change the top of the 'History' to a 'search/filter commits' textbox
- to the right of that, add a button that allows selecting the branch to compare against

This could be aligned to (a potentially simplified version of) the 'choose base branch' in GitHub web's 'Compare' ([Ref](https://github.com/desktop/desktop/compare/release-3.3.6...development))

![image](https://github.com/desktop/desktop/assets/753891/def54f7f-a498-4d63-b5f8-cf3453fd8563)

![image](https://github.com/desktop/desktop/assets/753891/daf8b685-958d-4d82-920d-7f6e1b258e88)

Or the 'Choose a base branch' on a GitHub web Pull Request.

![image](https://github.com/desktop/desktop/assets/753891/9442129d-a4ff-4fff-bbfe-6a2933a1e73f)

---

RE: https://github.com/desktop/desktop/issues/7022#issuecomment-1530008835

> Apparently this is a feature that was worked on at one point but it conflicted with some other features so it hasn't been included in the app yet and is on hold.
> 
> [`search-commits-by-text`](https://github.com/desktop/desktop/tree/search-commits-by-text?rgh-link-date=2023-05-01T17%3A52%3A05Z)

You can see the compare view here:

- https://github.com/desktop/desktop/compare/development...search-commits-by-text

Currently there is only a single commit, which you can see all the changed files/etc:

- https://github.com/desktop/desktop/commit/43819cfa92841a6da0905b993666a790c31853fa

**Comment by emmadesilva** [comment]
I've never felt the need to compare across branches, but filtering commits and searching for them is something I've often wanted. I think replacing the branch filter with something like this would be even better. Typing in it makes so that only search matches are shown in the history below.

![image](https://github.com/desktop/desktop/assets/95144705/a826f343-ce22-4e21-ae22-904a7e599367)

**Comment by LeftTwixWand** [comment]
Are there any updates?
Searching by the commit name is one of the most important features for me

**Comment by CliffracerX** [comment]
Waiting on this to be added as well, for filtering by author or *ideally* even by distinct commits in a branch (e.g, if a branch is 5 commits ahead of main, being able to filter to *just* those commits)

**Comment by NoxFfredrika** [comment]
I'm shocked that searching for a commit by name isn't already possible. Please add this feature.

**Comment by a2937** [comment]
I'd love to attempt this, only if there was a promise that my effort will not be in vain and someone would benefit from it.

**Comment by attheshow** [comment]
> I'd love to attempt this, only if there was a promise that my effort will not be in vain and someone would benefit from it.

I would definitely use it if there was a version that included the search feature. It's really the only feature that the current app is lacking in my opinion! 😊🏅

**Comment by emmadesilva** [comment]
> I'd love to attempt this, only if there was a promise that my effort will not be in vain and someone would benefit from it.

Would also really love this!

**Comment by NoxFfredrika** [comment]
I would use this feature a lot! Thanks for considering working on it!

**Comment by gingerbeardman** [comment]
Fingers crossed this makes it in "soon"

**Comment by attheshow** [comment]
If anyone here is a student, teacher, or part of a government non-profit you can apply for a free license for the Tower desktop app which includes search capabilities (apply at: https://www.git-tower.com/education/mac). I use Github Desktop for most of my work, but I unfortunately have to open Tower still for searching. I thought I'd recommend it for those who might be in a similar boat as me. 😊🏅

**Comment by Jedrzej94** [comment]
> If anyone here is a student, teacher, or part of a government non-profit you can apply for a free license for the Tower desktop app which includes search capabilities (apply at: https://www.git-tower.com/education/mac). I use Github Desktop for most of my work, but I unfortunately have to open Tower still for searching. I thought I'd recommend it for those who might be in a similar boat as me. 😊🏅

Shame I'm just a dev.

Anyway, does anyone work on this? @a2937 did you attempt or created a fork or something?

**Comment by ctrl-alt-lalit** [comment]
The lack of commit filtering has been a real bugbear for me too. At my day job, I've been using Desktop for most Git workflows, then GitExtensions for code review in a monorepo. I'd like to just use Desktop if possible.

I'd be interested in getting this to work. It took an evening for me to get some basic author filtering up and running. From a pure programming perspective, this isn't too hard -- just pass in some CLI flags to `git log` depending on what filters the user wants. Though I think getting consensus on what the UI & feature set should be is more difficult.

@sergiou87  or @steveward (or secret third Desktop programmer), what's the protocol for external contributions on larger features like this? Is there some sort of design template or process I need to follow?

**Comment by Seb33300** [comment]
For the UI, I would suggest to resuse the existing search input in the History tab.
Just make the left icon clickable to choose the search mode `Select branch` / `Search commit`
![image](https://github.com/user-attachments/assets/7d3e2c45-69b1-429d-9cbe-89077820c71c)

**Comment by sarim** [comment]
@ctrl-alt-lalit you can see there is a "not planned" tag on this issue. I don't see the comment here right now (or my bad eyes couldn't find it at first glance) but I remember something from the team in line of "community shouldn't start coding before we do a meeting and decide if and how we will add this feature". Seeing "not planned" tag on this now, seems like official team decided not to add this feature.... :(

**Comment by Jedrzej94** [comment]
> @ctrl-alt-lalit you can see there is a "not planned" tag on this issue. I don't see the comment here right now (or my bad eyes couldn't find it at first glance) but I remember something from the team in line of "community shouldn't start coding before we do a meeting and decide if and how we will add this feature". Seeing "not planned" tag on this now, seems like official team decided not to add this feature.... :(

Which would be dumb as box of rocks. This feature seems like is one of the most anticipated features yet, it's so basic yet it isn't there. Strange decisions strike again.

Perhaps it would be better to move onto some other software instead, which provides more functionalities.. (but gotta pay for them, ugh)

**Comment by emmadesilva** [comment]
> @ctrl-alt-lalit you can see there is a "not planned" tag on this issue. I don't see the comment here right now (or my bad eyes couldn't find it at first glance) but I remember something from the team in line of "community shouldn't start coding before we do a meeting and decide if and how we will add this feature". Seeing "not planned" tag on this now, seems like official team decided not to add this feature.... :(

That tag is often added to stale issues. Don't read into it too much.

**Comment by caiocinel** [comment]
+1

**Comment by collinph** [comment]
Come on guys. Do you not see the need--- and I mean the absolutely necessary, dire, and currently debilitating lack of a search feature? Did you guys write reddit or something? Search absolutely sucks, and it kills the ability to use this software by itself. Please tell me what the major malfunction is, and I'll be happy to do the work myself. I'm the guy who pushes your software in my company and it's outright embarrassing, especially since you can do what is needed from the CLI.

GitBox (for Mac) has had this for an eternity, and it's fast. Please provide an explanation of why this issue does not matter to the DEVs.. We need to search by file name's affected, authors, or other means. It can all be done within the search box already provided with no other UI necessary, it just needs to search better than it does.

**Comment by a2937** [comment]
> Anyway, does anyone work on this? @a2937 did you attempt or created a fork or something?

I created a fork and can presumably program the functionality if I had the time and some way of confirming that the Desktop moderators won't just close my PR. 

I know what's it like to see spam PRs and features that aren't wanted. I need to know if the staff want this. I know it can help the product.

**Comment by NopeUniversity** [comment]
As someone just now getting into using Git, this seems like a massive design oversight. There's like three different ways to check different branches, but there's no way other than manually scrolling and clicking through commits to find the one with the same hash you're looking for. For dealing with any project that has more than around 100 commits, or with more than 3-5 authors, this becomes incredibly burdensome and even downright hostile to use. I get that more experienced users of Git use the command line, but having a desktop client and omitting such a basic function as searching for a commit based on the hash raises the question of why the desktop client even exists. It's a tool designed to make life easier for version control that requires switching to a second tool for basic functionality. Is there a way to escalate this issue beyond this one issues board?

**Comment by HDaoud** [comment]
has anyone used this? does it solve the issue?
https://github.com/desktop/desktop/commit/43819cfa92841a6da0905b993666a790c31853fa

**Comment by Jedrzej94** [comment]
Seems promising but it was 4 years ago. Should cherry-pick that commit on the latest branch and check it out ourselves tbf.

**Comment by ipkpjersi** [comment]
The issue is, that commit seems to replace/remove some functionality, and I'm pretty sure they'd never sign off on that when they likely won't sign off on it even if it didn't replace/remove anything. Then again since they won't sign off either way I guess it doesn't matter lol

**Comment by Jedrzej94** [comment]
That is true, that it removes the previous feature. But OTOH, it doesn't mean we shouldn't try it (locally rebuild it with that commit).

Because if it does work. Then we have a good base to restore the previous (removed in that commit) functionality and add a switch/setting to use a "find commit by name" feature instead.

**Comment by Jedrzej94** [comment]
(that's a GIF, play it)

![Image](https://github.com/user-attachments/assets/5d3019ed-767a-48b7-958e-6170914e4f5a)

**Comment by ghnp5** [comment]
Thanks! That makes the Search field useful, and I'd actually start using it.

Comparing branches should be an option that I select from the top menus. It doesn't need to be a Search field that is always showing.

But searching commits... I'd give it so much use...!!

**Comment by Jedrzej94** [comment]
> Thanks! That makes the Search field useful, and I'd actually start using it.
> 
> Comparing branches should be an option that I select from the top menus. It doesn't need to be a Search field that is always showing.
> 
> But searching commits... I'd give it so much use...!!

I am willing to work on the double feature, where you can compare branches and search commits by commit name, but it totally depends if the team is actually eager to accept such feature.

So far, I can see a tag "not-planned" and that kinda speaks for itself, so it's hard to tell if I should actually continue at all.

**Comment by ghnp5** [comment]
Yeah, as far as I remember, the team wasn't much willing to have this done... and I can understand it would be a waste of time to develop something, and then it not to be accepted. Is there any way to ask/tag them, to see what they say about it?

**Comment by Seb33300** [comment]
> So far, I can see a tag "not-planned" and that kinda speaks for itself, so it's hard to tell if I should actually continue at all.

I think `not-planned` means there is no plan for a member of the core team to work on that feature yet.

But they are probably open for external contribution.

**Comment by sarim** [comment]
> > So far, I can see a tag "not-planned" and that kinda speaks for itself, so it's hard to tell if I should actually continue at all.
> 
> I think `not-planned` means there is no plan for a member of the core team to work on that feature yet.
> 
> But they are probably open for external contribution.

Please read the whole thread. They explicitly said not to work on this because they won't accept it. The feature needs to be okeyed by their "corporate" first.

**Comment by Seb33300** [comment]
@sarim, please share the message from a core team member saying they will not accept any external contribution about this?

I am following this thread since the beginning and I did not see such message.

**Comment by attheshow** [comment]
@Jedrzej94 Is there any way for us to use the version that you showed in your GIF? Do you have any instructions on how to compile that version so that we can test it out and play with it? Honestly I'm more interested in having this feature than most of the other features they've been adding in past years. So even if I have to start using a forked version, I'll probably be happier overall with the app. 😊

**Comment by Jedrzej94** [comment]
> [@Jedrzej94](https://github.com/Jedrzej94) Is there any way for us to use the version that you showed in your GIF? Do you have any instructions on how to compile that version so that we can test it out and play with it? Honestly I'm more interested in having this feature than most of the other features they've been adding in past years. So even if I have to start using a forked version, I'll probably be happier overall with the app. 😊

It's my fork, unpublished branch. I just made it myself last night in like 4 hours 👀 

I was thinking the same, of using personal fork and just keep rebasing with the main branch and fix conflicts on the go.

The branch from 4 years ago is totally unusable.

**Comment by Andrej730** [comment]
I've currently started to use Sublime Merge for commit search as a workaround - it seems to cover everything and even more, could be a good source of inspiration for Github Desktop search.

See example for available search options below:
 
![Image](https://github.com/user-attachments/assets/f96deca0-8810-4483-b5dd-8e65fdd5c22e)

**Comment by drewolbrich** [comment]
I would also like to be able to search for a commit by commit hash in GitHub Desktop, much like
```
git show caeed609c
```
Thank you. 😀

**Comment by matthewj-rfpro** [comment]
I can't believe that I still have to open a console and type this.
`git log --grep="my text"`
This is such a basic feature :-( 
5+ years of discussion? What's going on?

**Comment by collinph** [comment]
Is github desktop basically dead, or do the developers not actually do much development work? I hate to sound rude, but search is probably the most basic feature people expect.

**Comment by KNKNKNL** [comment]
Really in need of this feature🙏.

**Comment by collinph** [comment]
I need a developer who's "in the know" to give an explanation of why this issue has never been worked on. Do we need more community contribution, do we need to describe the problem or solution better.. or are you not interested in having good search? Other tools (most) have this functionality for git and it makes this tool seem like a tool for amateurs. I'm trying to express the seriousness of this need without being rude, but it's hard to do so. I'm happy to put in as many details as you like and even discuss how this was done in other tools. We cannot continue to use Github Desktop without a working search, and without a development team who cares that search works well.

We need to be able to search commit history by more than comments and be able to search DIFFS to start with.

**Comment by gingerbeardman** [comment]
> I need a developer who's "in the know" to give an explanation of why this issue has never been worked on. 

The priorities of GitHub/Microsoft do not align with those of GitHub Desktop Users.

**Comment by rsodre** [comment]
If we exit too often to GitHub or a terminal to do *something*, that *something* should be a feature.
99% when I exit GitHub Desktop is to: search for a commit, check the [network graph](https://github.com/desktop/desktop/issues/1634), and make [empty commits](https://github.com/desktop/desktop/issues/14903).
All stalled or rejected. 😕

**Comment by 0xdevalias** [comment]
> > I need a developer who's "in the know" to give an explanation of why this issue has never been worked on.
> 
> The priorities of GitHub/Microsoft do not align with those of GitHub Desktop Users.

@collinph The [answer commonly given on other issues](https://github.com/search?q=repo%3Adesktop%2Fdesktop+%22small+team%22&type=issues&s=&o=desc) is that the GitHub Desktop team is a small team with limited resources; and that because of that, they need to prioritise what they are working on/etc.

They have also sometimes made comments about wanting to keep GitHub Desktop easy to use for newer users and not necessarily add 'all the features' (paraphrasing). I'm not saying this feature does/doesn't fall under that reasoning. This section of the `README.md` + what it links out to may be helpful in that regard:

- https://github.com/desktop/desktop#is-github-desktop-right-for-me-what-are-the-primary-areas-of-focus
  - https://github.com/desktop/desktop/blob/development/docs/process/what-is-desktop.md

**Comment by sergiou87** [comment]
Also requested in #21543 by @dev-faizankhan

> ### The feature request
> ### Problem
> Currently, there is no way to search or filter commits in the History tab. When working with repositories that have a large commit history, users must manually scroll through the commit list to find a specific commit. This makes it difficult to locate commits by message, author, or SHA without switching to the command line.
> 
> ### Proposed solution
> Add a search bar above the commit list in the History tab that allows users to search commits by:
> 
> * **Commit message** (summary and body)
> * **Author name or email**
> * **Commit SHA** (full or partial, minimum 4 characters)
> 
> The search would use `git log --grep` and `--author` for efficient server-side filtering rather than loading all commits into memory.
> 
> ### User Experience
> 1. User navigates to the History tab
> 2. A search input with a search icon appears above the commit list
> 3. User types a query (minimum 3 characters to trigger search)
> 4. The commit list updates to show only matching commits
> 5. Clearing the search input restores the full commit history
> 
> ### Additional Details
> * Search is case-insensitive
> * Input is debounced (300ms) to avoid excessive git operations
> * Search is only available in History mode, not in Compare mode
> * Results are limited to 100 commits for performance
> * A "No matching commits" message is shown when the search yields no results
> 
> ### Use Cases
> * Finding a specific bug fix by searching the commit message
> * Finding all commits by a particular contributor
> * Quickly jumping to a commit by its SHA from a PR or issue reference
> 
> ### Additional context
> * Search is only active in History mode, not in Compare mode
> * A "No matching commits" message is shown when no results are found
> * The search input follows the existing FancyTextBox UI pattern used by the
>   branch filter, so it is visually consistent with the rest of the app
> * I have a working implementation ready to submit as a PR if this feature
>   is accepted

**Comment by dev-faizankhan** [comment]
Hi 👋

I noticed this issue was referenced from my duplicate issue #21543.

I have already implemented a working prototype for commit search in the History tab (search by message, author, and SHA using git log --grep and --author with debouncing and result limits).

I would be happy to open a PR if maintainers think this approach aligns with the project direction. Please let me know if I should proceed or if there are design/architecture guidelines I should follow first.

Thanks!

<img width="817" height="453" alt="Image" src="https://github.com/user-attachments/assets/7b60e676-d7f9-450d-8124-0257f83593d5" />
<img width="805" height="393" alt="Image" src="https://github.com/user-attachments/assets/0b71b970-ba80-4bf2-999c-28c4c0f36a28" />

**Comment by collinph** [comment]
Perhaps the people interested in this should just fork this project so we trade the official in for something that's actually maintained.

**Comment by guplem** [comment]
In case you are willing to switch to a fork, this has already been implemented in [GitHub Desktop Plus](https://github.com/pol-rivero/github-desktop-plus). It adds the ability to search commits in the History tab by title, message, tag, or hash.

**Comment by ghnp5** [comment]
I'm using that fork since a few weeks ago, and it's ABSOLUTELY GREAT, and works for Linux (unlike GitHub Desktop).

Not only this is implemented, but also the "auto stage" feature was just recently implemented, for files that are marked as "conflict resolved", that prevented you from switching branches until you open a terminal and do `git add`. This was one of the most annoying problems with GitHub Desktop for me. No problem, Desktop Plus implemented it without any issues, as soon as I asked for it.....

https://github.com/pol-rivero/github-desktop-plus/issues/86 -- Desktop Plus implemented as soon as I asked for it

https://github.com/desktop/desktop/issues/8800 -- GitHub Desktop simply closed my issue... 🙄
